File: packages/pinjected-linter/src/pinjected_linter/rules/pinj006_async_injected_naming.py
"""PINJ006: Async injected naming rule."""

import ast

from ..models import Fix, Position, RuleContext, Severity
from ..utils.ast_utils import has_decorator
from .base import ASTRuleBase, BaseNodeVisitor


class PINJ006Visitor(BaseNodeVisitor):
    """Visitor for checking async @injected function naming."""

    def visit_AsyncFunctionDef(self, node):
        """Check async function definitions."""
        if has_decorator(node, "injected") and not node.name.startswith("a_"):
            fix = self._create_fix(node)
            self.add_violation(
                node,
                f"Async @injected function '{node.name}' must have 'a_' prefix. "
                f"This helps distinguish async functions in dependency injection.",
                suggestion=f"Rename to 'a_{node.name}'",
                fix=fix,
            )
        self.generic_visit(node)

    def _create_fix(self, node: ast.AsyncFunctionDef) -> Fix:
        """Create a fix to add the 'a_' prefix."""
        old_name = node.name
        new_name = f"a_{old_name}"

        # The position after 'async def '
        start_pos = Position(
            line=node.lineno,
            column=node.col_offset + 10,  # 'async def ' is 10 characters
        )
        end_pos = Position(
            line=node.lineno,
            column=node.col_offset + 10 + len(old_name),
        )

        return Fix(
            start_pos=start_pos,
            end_pos=end_pos,
            replacement=new_name,
            description=f"Add 'a_' prefix to '{old_name}'",
        )


class PINJ006AsyncInjectedNaming(ASTRuleBase):
    """Rule for checking async @injected function naming convention.
    
    Async @injected functions MUST have 'a_' prefix to clearly indicate
    that they are asynchronous operations in Pinjected's dependency
    injection system.
    """
    
    rule_id = "PINJ006"
    name = "Async injected naming"
    description = "Async @injected functions must have 'a_' prefix"
    severity = Severity.ERROR
    category = "naming"
    auto_fixable = True
    
    def get_visitor(self, context: RuleContext) -> ast.NodeVisitor:
        """Get the AST visitor for this rule."""
        return PINJ006Visitor(self, context)

File: packages/pinjected-linter/rust-poc/src/rules/pinj006_async_injected_naming.rs
//! PINJ006: Async injected naming
//!
//! Async @injected functions must have 'a_' prefix.
//! This helps distinguish async functions in dependency injection.

use crate::models::{RuleContext, Severity, Violation};
use crate::rules::base::LintRule;
use crate::utils::pinjected_patterns::{has_async_prefix, has_injected_decorator_async};
use rustpython_ast::Stmt;

pub struct AsyncInjectedNamingRule;

impl AsyncInjectedNamingRule {
    pub fn new() -> Self {
        Self
    }
}

impl LintRule for AsyncInjectedNamingRule {
    fn rule_id(&self) -> &str {
        "PINJ006"
    }

    fn description(&self) -> &str {
        "Async @injected functions must have 'a_' prefix"
    }

    fn check(&self, context: &RuleContext) -> Vec<Violation> {
        let mut violations = Vec::new();

        match context.stmt {
            Stmt::AsyncFunctionDef(func) => {
                if has_injected_decorator_async(func) && !has_async_prefix(&func.name) {
                    violations.push(Violation {
                        rule_id: self.rule_id().to_string(),
                        message: format!(
                            "Async @injected function '{}' must have 'a_' prefix. \
                            This helps distinguish async functions in dependency injection.",
                            func.name
                        ),
                        offset: func.range.start().to_usize(),
                        file_path: context.file_path.to_string(),
                        severity: Severity::Error,
                    });
                }
            }
            _ => {}
        }

        violations
    }
}
